## Упражнения по библиотеке Numpy

In [2]:
import numpy as np

array = np.random.randint(1, 50, size=10)
print("Random array:", array)

Random array: [41 36 42  5 34  6 25 34 38 28]


**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [ ]:
# your code

array[(array>=3)&(array<=8)]*=-1
print("Array after multiplying elements in range [3, 8] by -1:", array)


Modified array: [-4 44 -3 28 38 46 38 42 19  2]


**2.** Заменить максимальный элемент случайного массива на 0

In [7]:
# your code

array[array.argmax()] = 0
print("Array after replacing max with 0:", array)


Array after replacing max with 0: [-4 44 -3 28 38  0 38 42 19  2]


**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [8]:
# your code

array2=np.array([[5,9],[1,3],[8,6]])
grid=np.meshgrid(array2[:,0], array2[:,1],indexing='ij')
result=np.column_stack([grid[0].ravel(), grid[1].ravel()])
print("Resulting array of all combinations of elements from the two columns:\n", result)


Resulting array of all combinations of elements from the two columns:
 [[5 9]
 [5 3]
 [5 6]
 [1 9]
 [1 3]
 [1 6]
 [8 9]
 [8 3]
 [8 6]]


**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [11]:
# your code

A = np.array([
    [1, 2, 3],
    [7, 8, 9],
    [3, 4, 5],
    [1, 1, 6]
])

B = np.array([
    [1, 3],
    [4, 6]
])

Array = A[np.all([np.isin(A, b_row).any(axis=1) for b_row in B], axis=0)]
print(Array)

[[3 4 5]
 [1 1 6]]


**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [ ]:
# your code

from numpy import matrix

matrix = np.array([[1, 2, 3], [4, 5, 6], [1, 2, 3], [7, 8, 9], [4, 5, 6], [1, 1, 1], [2, 2, 2], [3, 3, 3], [4, 4, 4], [5, 5, 5]])
result = matrix[matrix.max(axis=1) != matrix.min(axis=1)]
print("Matrix with unique elements within a row:")
print(result)


Matrix with unique rows:
[[1 2 3]
 [4 5 6]
 [1 2 3]
 [7 8 9]
 [4 5 6]]


**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [5]:
# your code

Array = np.array([
    [7, 8],
    [1, 2],
    [4, 5],
    [1, 2]  
    ])

k, indices = np.unique(Array, axis=0, return_index=True)
unique_arr = Array[np.sort(indices)]
print("Array without duplicate rows:")
print(unique_arr)

Array without duplicate rows:
[[7 8]
 [1 2]
 [4 5]]


______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [ ]:
# your code
 
import math 

Matrix = [[1,2,3], [4,5,6], [7,8,9]]
print("Product of non-zero diagonal elements:", math.prod([Matrix[i][i] for i in range(min(len(Matrix), len(Matrix[0]))) if Matrix[i][i] != 0]))


Product of non-zero diagonal elements: 45


In [16]:
import numpy as np

Matrix = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
print("Product of non-zero diagonal elements:", np.prod(np.diagonal(Matrix)[np.diagonal(Matrix) != 0]))

Product of non-zero diagonal elements: 45


* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [24]:
# your code

x = [1, 2, 2, 4]
y = [4, 2, 1, 2]

print(sorted(x) == sorted(y))


True


In [25]:
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])

print(np.array_equal(np.sort(x), np.sort(y)))

True


* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [29]:
# your code

x = [6, 2, 0, 3, 0, 0, 5, 7, 0]

next_ind0 = [x[i + 1] for i in range(len(x) - 1) if x[i] == 0]
print(max(next_ind0))

5


In [28]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])

ind0 = np.where(x == 0)[0]
ind0 = ind0[ind0 < len(x) - 1]  
next_ind0 = x[ind0 + 1]
print(next_ind0.max())

5


* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [ ]:
# your code

x = [2, 2, 2, 3, 3, 3, 5]

values = []
counts = []
for i in x:
    if not values or i != values[-1]:
        values.append(i)
        counts.append(1)
    else:
        counts[-1] += 1
print(values, counts)


[2, 3, 5] [3, 3, 1]


In [ ]:

x = np.array([2, 2, 2, 3, 3, 3, 5])

groups = np.split(x, np.where(np.diff(x) != 0)[0] + 1)
values = np.array([g[0] for g in groups])
counts = np.array([len(g) for g in groups])
print(values, counts)

[2 3 5] [3 3 1]


* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [39]:
!pip install scipy
from scipy.spatial.distance import cdist
import time

In [40]:
# your code

np.random.seed(42)
X = np.random.rand(1000, 10)
Y = np.random.rand(500, 10)

start_time = time.time()
res_scipy = cdist(X, Y, metric='euclidean')
time_scipy = time.time() - start_time
print(f"Scipy time: {time_scipy}")

Scipy time: 0.005307912826538086


In [41]:

np.random.seed(42)
X = np.random.rand(1000, 10)
Y = np.random.rand(500, 10)

start_time = time.time()
x_sq = np.sum(X**2, axis=1, keepdims=True)
y_sq = np.sum(Y**2, axis=1, keepdims=True).T
cross_term = np.dot(X, Y.T)
time_numpy = time.time() - start_time
print(f"NumPy time: {time_numpy:.5f} сек.")

NumPy time: 0.04067 сек.


_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [42]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [43]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [46]:
# your code


mean_calories_stats = np.mean(calorie_stats)
my_calories = 60
average_calories = mean_calories_stats-my_calories
print("Average calories difference:", average_calories)

Average calories difference: 46.883116883116884


3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [48]:
# your code

calorie_stats_sorted = np.sort(calorie_stats)
print("Sorted calorie stats:", calorie_stats_sorted)

Sorted calorie stats: [ 50.  50.  50.  70.  70.  80.  90.  90.  90.  90.  90.  90.  90. 100.
 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100.
 100. 100. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 120. 120. 120. 120. 120. 120. 120. 120. 120. 120. 130.
 130. 140. 140. 140. 150. 150. 160.]


4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [ ]:
# your code

median_calorie_stats = np.median(calorie_stats)
print("Median calorie stats:", median_calorie_stats)

Median calorie stats: 110.0


5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [59]:
# your code

print("75th percentile of calorie stats:", np.percentile(calorie_stats, 75))
print("50th percentile of calorie stats:", np.percentile(calorie_stats, 50))
print("25th percentile of calorie stats:", np.percentile(calorie_stats, 25))
print("10th percentile of calorie stats:", np.percentile(calorie_stats, 10))
print("7th percentile of calorie stats:", np.percentile(calorie_stats, 7))
print("5th percentile of calorie stats:", np.percentile(calorie_stats, 5))
print("3rd percentile of calorie stats:", np.percentile(calorie_stats, 3))
print("4th percentile of calorie stats:", np.percentile(calorie_stats, 4))

nth_percentile = 4
print(f"{nth_percentile}th percentile of calorie stats:", np.percentile(calorie_stats, nth_percentile))


75th percentile of calorie stats: 110.0
50th percentile of calorie stats: 110.0
25th percentile of calorie stats: 100.0
10th percentile of calorie stats: 90.0
7th percentile of calorie stats: 83.2
5th percentile of calorie stats: 70.0
3rd percentile of calorie stats: 55.599999999999994
4th percentile of calorie stats: 70.0
4th percentile of calorie stats: 70.0


6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [60]:
# your code

more_calories = (sum(1 for cal in calorie_stats if cal > 60) / len(calorie_stats)) * 100
print("Percentage of cereals with more than 60 calories:", more_calories)

Percentage of cereals with more than 60 calories: 96.1038961038961


7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [ ]:
# your code

calorie_std = np.std(calorie_stats)
print("Standard deviation of calorie stats:", calorie_std)

#The mean difference of 46.8 calories shows a substantial gap between categories, but the standard deviation of 19.3 reveals considerable variation within the data itself

Standard deviation of calorie stats: 19.35718533390827


8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Analysis shows that CrunchieMunchies with 60 calories is a much lighter choice compared to competitors averaging 106.8 calories. Since 96.1 percent of other cereals have a higher calorie count, our product is clearly the healthier option for consumers. Even though the standard deviation of 19.4 shows some variation across brands, the data proves that CrunchieMunchies consistently outperforms competitors in offering a lower-calorie product for a healthy lifestyle.